go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,297 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encoding
+
+import (
+	"math"
+
+	"github.com/apache/arrow/go/v17/arrow"
+	"github.com/apache/arrow/go/v17/internal/bitutils"
+	"github.com/apache/arrow/go/v17/internal/utils"
+	"github.com/apache/arrow/go/v17/parquet"
+	"golang.org/x/xerrors"
+)
+
+type NumericByteStreamSplitType interface {
+	float32 | float64 | int32 | int64
+}
+
+type ByteStreamSplitType interface {
+	NumericByteStreamSplitType | parquet.FixedLenByteArray
+}
+
+// putByteStreamSplitNumeric is a generic implementation of the BYTE_STREAM_SPLIT encoding for fixed-width numeric types.
+func putByteStreamSplitNumeric[T NumericByteStreamSplitType](in []T, enc TypedEncoder, sink *PooledBufferWriter) {
+	numElements := len(in)
+	typeLen := enc.Type().ByteSize()
+	bytesNeeded := numElements * typeLen
+	sink.Reserve(bytesNeeded)
+
+	// Make sure len = cap so we can index into any loc rather than append
+	data := sink.buf.Bytes()
+	data = data[:cap(data)]
+
+	inBytes := arrow.GetBytes(in)
+	for offset := 0; offset < typeLen; offset++ {
+		for element := range in {
+			encLoc := numElements*offset + element
+			decLoc := typeLen*element + offset
+			data[encLoc] = inBytes[decLoc]
+		}
+	}
+
+	sink.pos += bytesNeeded
+}
+
+// putByteStreamSplitSpaced encodes data that has space for nulls using the BYTE_STREAM_SPLIT encoding, calling to the provided putFn to encode runs of non-null values.
+func putByteStreamSplitSpaced[T ByteStreamSplitType](in []T, validBits []byte, validBitsOffset int64, bitSetReader bitutils.SetBitRunReader, putFn func([]T)) {
+	if validBits == nil {
+		putFn(in)
+		return
+	}
+
+	if bitSetReader == nil {
+		bitSetReader = bitutils.NewSetBitRunReader(validBits, validBitsOffset, int64(len(in)))
+	} else {
+		bitSetReader.Reset(validBits, validBitsOffset, int64(len(in)))
+	}
+
+	for {
+		run := bitSetReader.NextRun()
+		if run.Length == 0 {
+			break
+		}
+		putFn(in[int(run.Pos):int(run.Pos+run.Length)])
+	}
+}
+
+// decodeByteStreamSplit is a generic implementation of the BYTE_STREAM_SPLIT decoder for fixed-width numeric types.
+func decodeByteStreamSplitNumeric[T NumericByteStreamSplitType](out []T, dec TypedDecoder, data []byte) (int, error) {
+	max := utils.Min(len(out), dec.ValuesLeft())
+	typeLen := dec.Type().ByteSize()
+	numBytesNeeded := max * typeLen
+	if numBytesNeeded > len(data) || numBytesNeeded > math.MaxInt32 {
+		return 0, xerrors.New("parquet: eof exception")
+	}
+
+	outBytes := arrow.GetBytes(out)
+	for offset := 0; offset < typeLen; offset++ {
+		for element := 0; element < max; element++ {
+			encLoc := max*offset + element
+			decLoc := typeLen*element + offset
+			outBytes[decLoc] = data[encLoc]
+		}
+	}
+
+	dec.SetData(dec.ValuesLeft()-max, data[max:])
```

Comment:
Hmmm so `dec` handles this by `SetData` repeatly?

Reply:
All `SetData` is doing here is updating the position/stats for the decoder, so that it will start again from the right place in subsequent calls. Normally decoders increment the `pos` one at a time, but since we are decoding a fixed size chunk of size K x N we just move the `pos` by that much before returning.